server/etcdserver/server.go

```diff
@@ -858,8 +858,20 @@ func (s *EtcdServer) run() {
 }
 
 func (s *EtcdServer) revokeExpiredLeases(leases []*lease.Lease) {
+	lg := s.Logger()
+	newLeaderId := s.raftStatus().Lead
```

Comment:
Is `s.raftStatus().Lead` updated if the raft loop is blocked? I was thinking that we need to make a quorum read here to confirm leadership.

Reply:
> Is s.raftStatus().Lead updated if the raft loop is blocked? 

No, because it creates a separate channel to receive the response (raftStatus). FYI. https://github.com/etcd-io/raft/blob/7c3ed830bbb092753012c12fca1b7d4d29095552/node.go#L570

> I was thinking that we need to make a quorum read here to confirm leadership.

Already added `linearizableReadNotify` check. The new method `ensureLeadership` should be enough to guarantee the current member's leadership.

Reply:
> Is s.raftStatus().Lead updated if the raft loop is blocked?
>
> No, because it creates a separate channel to receive the response (raftStatus). FYI. https://github.com/etcd-io/raft/blob/7c3ed830bbb092753012c12fca1b7d4d29095552/node.go#L570

Ok, raftStatus is not blocked. But will it serve latest raft state or will it be outdated?

Reply:
> But will it serve latest raft state or will it be outdated?

Pls see line 906, `linearizableReadNotify` guarantees that the raft state is latest.

Reply:
Ok, was just confirming how `raftStatus()` works on it's own. Possibly we should consider looking through all calls to `raftStatus` and double checking if they will benefit from `linearizableReadNotify` before it.

Reply:
> Possibly we should consider looking through all calls to `raftStatus` and double checking if they will benefit from `linearizableReadNotify` before it.

It's a good point. Pls see https://github.com/etcd-io/etcd/issues/15944#issuecomment-1778860437

Reply:
> Ok, raftStatus is not blocked. But will it serve latest raft state or will it be outdated?

It could depend on in which failure mode. If it is just raft loop blocked while network is intact, the old leader will step down to follower after receiving a message higher than current term from new leader. Hence, raftStatus is updated while etcd server is not. 

If leader is both network partitioned and raft loop blocked, then raftStatus is outdated unless we separate the heartbeat ticker to a separate for loop.